content/blog/2021/11/deep-dive-into-policy-logging.md

```diff
@@ -0,0 +1,221 @@
+logging events to the host by using
+waPC](https://github.com/kubewarden/policy-sdk-rust/blob/642d13b07053be6455c0b5f49df0e87ff6f022c5/src/logging/drain.rs#L57-L60).
+
+You can read more about slog [here](https://github.com/slog-rs/slog).
+
+### Swift
+
+We will be looking at the [Swift policy
+template](https://github.com/kubewarden/swift-policy-template) for this example.
+
+As happens with Go and Rust's SDK's, the Swift SDK is instrumented to use Swift's `LogHandler`, so
+our policy only has to initialize it. In our `Sources/Policy/main.swift` file:
+
+```swift
+import kubewardenSdk
+import Logging
+
+LoggingSystem.bootstrap(PolicyLogHandler.init)
+```
+
+Then, in our policy business logic, under `Sources/BusinessLogic/validate.swift`, we are able to log
+with [different levels](https://apple.github.io/swift-log/docs/current/Logging/Structs/Logger.html):
+
+```swift
+import Logging
+
+public func validate(payload: String) -> String {
+    // ...
+
+    logger.info("validating object",
+        metadata: [
+            "some-key": "some-value",
+        ])
+
+    // ...
+}
+```
+
+Following the same strategy as the Go and Rust SDK's, the Swift SDK is able to [push log events to
+the host through
+waPC](https://github.com/kubewarden/policy-sdk-swift/blob/59cc979fa9994f1653d3d3e3dd3072188b2c0a18/Sources/kubewardenSdk/logger.swift#L75-L80).
+
+## Cluster Administrators
+
+Being able to log from within a policy is half of the story. Then, we have to be able to read and
+potentially collect these logs.
+
+As we have seen, Kubewarden policies support structured logging that is then forwarded to the
+component running the policy. Usually, this is `kwctl` if you are executing the policy in a manual
+fashion, or the `policy-server` if the policy is being ran in a Kubernetes environment.
+
+Both `kwctl` and the  `policy-server` use the `tracing` crate to produce log events, either the
+events that are produced by the application itself, or by policies that are running in WebAssembly
+runtime environments.
+
+### `kwctl`
+
+The `kwctl` CLI tool takes a very straightforward approach to logging from policies: it will print
+them to the standard error file descriptor.
+
+### `policy-server`
+
+The `policy-server` supports [different log
+formats](https://github.com/kubewarden/policy-server/blob/d615bfe7fdf1fe6001e655360fe85f0db2194410/src/cli.rs#L174-L206):
+`json`, `text` and `otlp`.
+
+`otlp`? I hear you ask. It stands for OpenTelemetry Protocol. We will look into that in a bit.
+
+If the `policy-server` is run with the `--log-fmt` argument set to `json` or `text`, the output will
+be printed to the standard error file descriptor in JSON or plain text formats.
+
+If `--log-fmt` is set to `otlp`, the `policy-server` will use OpenTelemetry to report logs and
+traces.
+
+## OpenTelemetry
+
+Kubewarden is instrumented with OpenTelemetry, so it's possible for the `policy-server` to send
```

Comment:
```suggestion
Kubewarden is instrumented with [OpenTelemetry](https://opentelemetry.io/), so it's possible for the `policy-server` to send
```

Reply:
We added a link on the first occurence of the term. Should I keep adding links to it?

Reply:
No we shouldn't. Sorry, I missed the first link